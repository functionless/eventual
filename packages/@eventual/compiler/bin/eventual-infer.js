#!/usr/bin/env -S node -r "ts-node/register"

const { infer } = require("../lib/cjs/eventual-infer.js");

infer();
