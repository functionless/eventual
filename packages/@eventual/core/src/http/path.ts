export type ParsePath<
  Text extends string,
  FoundNames extends string = never
> = Text extends `:${infer NameHead extends Word}${infer Rest}`
  ? TakeWhile<Rest, Word> extends [
      infer Tail extends string,
      infer NameTail extends string
    ]
    ? ParsePath<Tail, FoundNames | `${NameHead}${NameTail}`>
    : FoundNames
  : Text extends `${Word | "/"}${infer Rest}`
  ? ParsePath<Rest, FoundNames>
  : FoundNames;

type TakeWhile<
  Input extends string,
  Match extends string,
  Result extends string = ""
> = Input extends `${infer C}${infer Rest}`
  ? C extends Match
    ? TakeWhile<Rest, Match, `${Result}${C}`>
    : [Input, Result]
  : [Input, Result];

type Digit = "0" | "1" | "2" | "3" | "4" | "5" | "6" | "7" | "8" | "9";

type Word = Alphanumeric | "-" | "_";
type Alphanumeric = Digit | Letter;
type Letter = UppercaseLetter | LowercaseLetter;
type UppercaseLetter = Uppercase<LowercaseLetter>;
type LowercaseLetter =
  | "a"
  | "b"
  | "c"
  | "d"
  | "e"
  | "f"
  | "g"
  | "h"
  | "i"
  | "j"
  | "k"
  | "l"
  | "m"
  | "n"
  | "o"
  | "p"
  | "q"
  | "r"
  | "s"
  | "t"
  | "u"
  | "v"
  | "w"
  | "x"
  | "y"
  | "z";
