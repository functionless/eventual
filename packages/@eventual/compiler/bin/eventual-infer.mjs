#!/usr/bin/env node

import { infer } from "../lib/cjs/eventual-infer.js";

infer();
