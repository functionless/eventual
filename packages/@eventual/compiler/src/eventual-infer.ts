/**
 * This script imports a user's script and outputs a JSON object
 * to stdout containing all of the data that can be inferred.
 *
 * @see ServiceSpec
 */
import { generateSchema } from "@anatine/zod-openapi";
import {
  ServiceSpec,
  commands,
  entities,
  events,
  subscriptions,
  tasks,
  transactions,
  workflows,
  buckets,
  BucketStreamSpec,
} from "@eventual/core/internal";
import {
  CallExpression,
  ExportDeclaration,
  Expression,
  ModuleDeclaration,
  TsType,
  parseFile,
} from "@swc/core";
import { Visitor } from "@swc/core/Visitor.js";
import crypto from "crypto";
import esbuild from "esbuild";
import fs from "fs/promises";
import os from "os";
import path from "path";
import {
  getSpan,
  isCommandCall,
  isEntityStreamCall,
  isEntityStreamMemberCall,
  isOnEventCall,
  isSubscriptionCall,
  isTaskCall,
} from "./ast-util.js";
import { printModule } from "./print-module.js";

export async function infer(
  scriptName = process.argv[2],
  openApi: ServiceSpec["openApi"]
): Promise<ServiceSpec> {
  if (scriptName === undefined) {
    throw new Error(`scriptName undefined`);
  }

  const tmp = os.tmpdir();

  const bundle = await esbuild.build({
    mainFields: ["module", "main"],
    entryPoints: [scriptName],
    plugins: [inferPlugin],
    sourcemap: false,
    bundle: true,
    write: false,
    platform: "node",
  });

  const script = bundle.outputFiles[0]!.text;
  const hash = crypto.createHash("md5").update(script).digest("hex");
  scriptName = path.join(tmp, `${hash}.js`);
  await fs.writeFile(scriptName, script);

  await import(path.resolve(scriptName));

  const serviceSpec: ServiceSpec = inferFromMemory(openApi);

  console.log(JSON.stringify(serviceSpec));

  return serviceSpec;
}

export function inferFromMemory(openApi: ServiceSpec["openApi"]): ServiceSpec {
  return {
    workflows: [...workflows().keys()].map((n) => ({ name: n })),
    tasks: Object.values(tasks()).map((task) => ({
      name: task.name,
      sourceLocation: task.sourceLocation,
      options: task.options,
    })),
    events: Array.from(events().values()).map((event) => ({
      name: event.name,
      schema: event.schema ? generateSchema(event.schema) : undefined,
    })),
    subscriptions: subscriptions().map((e) => ({
      name: e.name,
      props: {
        memorySize: e.props?.memorySize,
        retryAttempts: e.props?.retryAttempts,
        handlerTimeout: e.props?.handlerTimeout,
      },
      sourceLocation: e.sourceLocation,
      filters: e.filters,
    })),
    commands: commands.map((command) => ({
      name: command.name,
      sourceLocation: command.sourceLocation,
      path: command.path,
      description: command.description,
      summary: command.summary,
      memorySize: command.memorySize,
      handlerTimeout: command.handlerTimeout,
      method: command.method,
      input: command.input ? generateSchema(command.input) : undefined,
      output: command.output ? generateSchema(command.output) : undefined,
      passThrough: command.passThrough,
      params: command.params,
      validate: command.validate,
      namespace: command.namespace,
    })),
    buckets: {
      buckets: [...buckets().values()].map((b) => ({
        name: b.name,
        streams: b.streams.map(
          (s) =>
            ({
              name: s.name,
              bucketName: s.bucketName,
              options: s.options,
              sourceLocation: s.sourceLocation,
            } satisfies BucketStreamSpec)
        ),
      })),
    },
    entities: {
      entities: [...entities().values()].map((d) => ({
        name: d.name,
        schema: d.schema ? generateSchema(d.schema) : undefined,
        streams: d.streams.map((s) => ({
          name: s.name,
          entityName: s.entityName,
          options: s.options,
          sourceLocation: s.sourceLocation,
        })),
      })),
    },
    transactions: [...transactions().values()].map((t) => ({
      name: t.name,
    })),
    openApi,
  };
}

export const inferPlugin: esbuild.Plugin = {
  name: "eventual",
  setup(build) {
    build.onLoad({ filter: /\.[mc]?[tj]s$/g }, async (args) => {
      // FYI: SWC erases comments: https://github.com/swc-project/swc/issues/6403
      const sourceModule = await parseFile(args.path, {
        syntax: "typescript",
      });

      const inferVisitor = new InferVisitor(args.path);
      const transformedModule = inferVisitor.visitModule(sourceModule);

      if (inferVisitor.didMutate) {
        const { code } = await printModule(transformedModule, args.path);

        return {
          contents: code,
          loader: "js",
        };
      }
      return undefined;
    });
  },
};

export class InferVisitor extends Visitor {
  public didMutate: boolean = false;

  private exportName: string | undefined;

  constructor(readonly fileName: string) {
    super();
  }

  public visitTsType(n: TsType): TsType {
    return n;
  }

  visitExportDeclaration(decl: ExportDeclaration): ModuleDeclaration {
    if (decl.declaration.type === "VariableDeclaration") {
      if (
        !decl.declaration.declare &&
        decl.declaration.declarations.length === 1
      ) {
        const varDecl = decl.declaration.declarations[0]!;
        if (varDecl.id.type === "Identifier") {
          if (varDecl.init?.type === "CallExpression") {
            this.exportName = varDecl.id.value;

            const call = this.visitCallExpression(varDecl.init);

            this.exportName = undefined;

            return {
              ...decl,
              declaration: {
                ...decl.declaration,
                declarations: [
                  {
                    ...varDecl,
                    init: call,
                  },
                ],
              },
            };
          }
        }
      }
    }
    return super.visitExportDeclaration(decl);
  }

  visitCallExpression(call: CallExpression): Expression {
    if (
      this.exportName &&
      (isCommandCall(call) ||
        isOnEventCall(call) ||
        isSubscriptionCall(call) ||
        isTaskCall(call) ||
        isEntityStreamMemberCall(call) ||
        isEntityStreamCall(call))
    ) {
      this.didMutate = true;

      return {
        ...call,
        arguments: [
          {
            expression: {
              type: "ObjectExpression",
              span: getSpan(call),
              properties: [
                {
                  type: "KeyValueProperty",
                  key: {
                    type: "Identifier",
                    optional: false,
                    span: getSpan(call),
                    value: "exportName",
                  },
                  value: {
                    type: "StringLiteral",
                    span: getSpan(call),
                    value: this.exportName,
                  },
                },
                {
                  type: "KeyValueProperty",
                  key: {
                    type: "Identifier",
                    optional: false,
                    span: getSpan(call),
                    value: "fileName",
                  },
                  value: {
                    type: "StringLiteral",
                    span: {
                      ctxt: 0,
                      end: 0,
                      start: 0,
                    },
                    value: this.fileName,
                  },
                },
              ],
            },
          },
          ...call.arguments,
        ],
      };
    }
    return super.visitCallExpression(call);
  }
}
