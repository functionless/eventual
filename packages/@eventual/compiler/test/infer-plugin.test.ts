import "jest";

import path from "path";
import esbuild from "esbuild";
import { inferPlugin } from "../src/eventual-infer.js";
import url from "url";

const __dirname = url.fileURLToPath(new URL(".", import.meta.url));

test("api handlers should be decorated with source location", async () => {
  const bundle = await esbuild.build({
    mainFields: ["module", "main"],
    entryPoints: [
      path.resolve(__dirname, "..", "test-files", "api-handler.ts"),
    ],
    sourcemap: false,
    plugins: [inferPlugin],
    bundle: true,
    write: false,
    tsconfig: "tsconfig.json",
    platform: "node",
  });

  expect(sanitizeBundle(bundle)).toMatchSnapshot();
});

function sanitizeBundle(
  bundle: esbuild.BuildResult & {
    outputFiles: esbuild.OutputFile[];
  }
) {
  return (
    bundle
      .outputFiles![0]?.text.split("\n")
      // HACK: filter out comment that is breaking the tests when run from VS Code
      // TODO: figure out why running vs code test is having trouble identifying the right
      //       tsconfig.test.json without a configuration at the root.
      // HINT: something to do with `.vscode/launch.json`
      .filter((line) => !(line.includes("fileName:") || line.includes("// ")))
      .join("\n")
  );
}
