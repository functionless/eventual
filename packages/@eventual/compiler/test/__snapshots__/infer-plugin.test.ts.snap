// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`api handlers should be decorated with source location 1`] = `
""use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

var command_worker_exports = {};
__export(command_worker_exports, {
  myHandler: () => myHandler,
  myHandler2: () => myHandler2
});
module.exports = __toCommonJS(command_worker_exports);

var CallKind;
(function(CallKind2) {
  CallKind2[CallKind2["AwaitTimerCall"] = 1] = "AwaitTimerCall";
  CallKind2[CallKind2["BucketCall"] = 10] = "BucketCall";
  CallKind2[CallKind2["ChildWorkflowCall"] = 7] = "ChildWorkflowCall";
  CallKind2[CallKind2["ConditionCall"] = 2] = "ConditionCall";
  CallKind2[CallKind2["EmitEventsCall"] = 4] = "EmitEventsCall";
  CallKind2[CallKind2["EntityCall"] = 8] = "EntityCall";
  CallKind2[CallKind2["ExpectSignalCall"] = 3] = "ExpectSignalCall";
  CallKind2[CallKind2["GetExecutionCall"] = 14] = "GetExecutionCall";
  CallKind2[CallKind2["InvokeTransactionCall"] = 9] = "InvokeTransactionCall";
  CallKind2[CallKind2["RegisterSignalHandlerCall"] = 5] = "RegisterSignalHandlerCall";
  CallKind2[CallKind2["SendSignalCall"] = 6] = "SendSignalCall";
  CallKind2[CallKind2["TaskCall"] = 0] = "TaskCall";
  CallKind2[CallKind2["TaskRequestCall"] = 12] = "TaskRequestCall";
  CallKind2[CallKind2["SearchCall"] = 11] = "SearchCall";
  CallKind2[CallKind2["StartWorkflowCall"] = 13] = "StartWorkflowCall";
})(CallKind || (CallKind = {}));

globalThis._eventual ??= { resources: {} };
function registerEventualResource(resourceKind, resource) {
  if (globalThis._eventual.resources[resourceKind]?.has(resource.name)) {
    throw new Error(\`\${resourceKind} with name '\${resource.name}' already exists\`);
  }
  (globalThis._eventual.resources[resourceKind] ??= /* @__PURE__ */ new Map()).set(resource.name, resource);
  return resource;
}

var PropertyKind;
(function(PropertyKind2) {
  PropertyKind2[PropertyKind2["BucketPhysicalName"] = 0] = "BucketPhysicalName";
  PropertyKind2[PropertyKind2["OpenSearchClient"] = 1] = "OpenSearchClient";
  PropertyKind2[PropertyKind2["ServiceClient"] = 2] = "ServiceClient";
  PropertyKind2[PropertyKind2["ServiceName"] = 3] = "ServiceName";
  PropertyKind2[PropertyKind2["ServiceSpec"] = 4] = "ServiceSpec";
  PropertyKind2[PropertyKind2["ServiceUrl"] = 5] = "ServiceUrl";
})(PropertyKind || (PropertyKind = {}));

function isSourceLocation(a) {
  return a && typeof a === "object" && typeof a.fileName === "string" && typeof a.exportName === "string";
}

var ServiceType;
(function(ServiceType2) {
  ServiceType2["CommandWorker"] = "CommandWorker";
  ServiceType2["Subscription"] = "Subscription";
  ServiceType2["OrchestratorWorker"] = "OrchestratorWorker";
  ServiceType2["EntityStreamWorker"] = "EntityStreamWorker";
  ServiceType2["BucketNotificationHandlerWorker"] = "BucketNotificationHandlerWorker";
  ServiceType2["TaskWorker"] = "TaskWorker";
  ServiceType2["TransactionWorker"] = "TransactionWorker";
})(ServiceType || (ServiceType = {}));

var SignalTargetType;
(function(SignalTargetType2) {
  SignalTargetType2[SignalTargetType2["Execution"] = 0] = "Execution";
  SignalTargetType2[SignalTargetType2["ChildExecution"] = 1] = "ChildExecution";
})(SignalTargetType || (SignalTargetType = {}));

var ExecutionStatus;
(function(ExecutionStatus2) {
  ExecutionStatus2["IN_PROGRESS"] = "IN_PROGRESS";
  ExecutionStatus2["SUCCEEDED"] = "SUCCEEDED";
  ExecutionStatus2["FAILED"] = "FAILED";
})(ExecutionStatus || (ExecutionStatus = {}));
var ExecutionStatuses = new Set(Object.values(ExecutionStatus));

function e({ base: t = "", routes: n = [] } = {}) {
  return { __proto__: new Proxy({}, { get: (e2, a, o) => (e3, ...r) => n.push([a.toUpperCase(), RegExp(\`^\${(t + e3).replace(/(\\/?)\\*/g, "($1.*)?").replace(/(\\/$)|((?<=\\/)\\/)/, "").replace(/:(\\w+)(\\?)?(\\.)?/g, "$2(?<$1>[^/]+)$2$3").replace(/\\.(?=[\\w(])/, "\\\\.").replace(/\\)\\.\\?\\(([^\\[]+)\\[\\^/g, "?)\\\\.?($1(?<=\\\\.)[^\\\\.")}/*$\`), r]) && o }), routes: n, async handle(e2, ...r) {
    let a, o, t2 = new URL(e2.url);
    e2.query = Object.fromEntries(t2.searchParams);
    for (var [p, s, u] of n)
      if ((p === e2.method || "ALL" === p) && (o = t2.pathname.match(s))) {
        e2.params = o.groups;
        for (var c of u)
          if (void 0 !== (a = await c(e2.proxy || e2, ...r)))
            return a;
      }
  } };
}
var itty_router_min_default = { Router: e };

function command(...args) {
  const [sourceLocation, name, options, handler] = parseCommandArgs(args);
  const command2 = {
    kind: "Command",
    name,
    handler,
    sourceLocation,
    ...options,
    output: options?.output ? "restStatusCode" in options.output ? options.output : { schema: options.output, description: "OK", restStatusCode: 200 } : { schema: void 0, description: "OK", restStatusCode: 200 }
  };
  return registerEventualResource("Command", command2);
}
function parseCommandArgs(args) {
  return [
    args.find(isSourceLocation),
    args.find((a) => typeof a === "string"),
    args.find((a) => typeof a === "object" && !isSourceLocation(a)),
    args.find((a) => typeof a === "function")
  ];
}

var router = itty_router_min_default.Router();
var api = createRouter([]);
function createRouter(middlewares) {
  return new Proxy({}, {
    get: (_, method) => {
      if (method === "routes" || method === "handle") {
        return router[method];
      } else if (method === "use") {
        return (middleware) => createRouter([...middlewares ?? [], middleware]);
      } else if (method === "command") {
        return (...args) => {
          const [sourceLocation, name, options, handler] = parseCommandArgs(args);
          return command(sourceLocation, name, {
            ...options ?? {},
            middlewares
          }, handler);
        };
      } else {
        return (...args) => {
          const [sourceLocation, path, routeProps, handler] = typeof args[0] === "object" ? typeof args[3] === "function" ? args : [
            args[0],
            args[1],
            void 0,
            args[2]
          ] : typeof args[2] === "function" ? [
            void 0,
            args[0],
            args[1],
            args[2]
          ] : [void 0, args[0], void 0, args[1]];
          const command2 = {
            description: routeProps?.description,
            kind: "Command",
            handler,
            memorySize: routeProps?.memorySize,
            method: method.toUpperCase(),
            name: path,
            path: typeof args[0] === "string" ? args[0] : args[1],
            sourceLocation,
            handlerTimeout: routeProps?.handlerTimeout,
            middlewares,
            otherOutputs: routeProps?.outputs,
            passThrough: true
          };
          registerEventualResource("Command", command2);
          return router[method](path, command2.handler);
        };
      }
    }
  });
}

var BaseHttpPayload = class {
  async tryJson() {
    try {
      return await this.json();
    } catch {
      return void 0;
    }
  }
  async json() {
    return JSON.parse(await this.text?.() ?? "");
  }
  async text() {
    if (this.body === void 0) {
      return "";
    } else if (typeof this.body === "string") {
      return this.body;
    } else if (Buffer.isBuffer(this.body)) {
      return this.body.toString("utf-8");
    } else {
      return Buffer.from((await readStream(this.body)).buffer).toString("utf-8");
    }
  }
  async arrayBuffer() {
    if (this.body === void 0) {
      return new ArrayBuffer(0);
    } else if (typeof this.body === "string") {
      return Buffer.from(this.body, "utf8");
    } else if (Buffer.isBuffer(this.body)) {
      return this.body;
    } else {
      return readStream(this.body);
    }
  }
};
var HttpResponse = class extends BaseHttpPayload {
  constructor(body, init) {
    super();
    this.body = body === void 0 ? null : body;
    this.status = init?.status ?? 200;
    this.statusText = init?.statusText;
    this.headers = toHeaders(init?.headers);
  }
};
function toHeaders(headers) {
  if (headers === void 0) {
    return new Headers();
  } else if (headers instanceof Headers) {
    return headers;
  } else {
    const h = new Headers();
    for (const [k, v] of Object.entries(headers)) {
      h.set(k, v);
    }
    return h;
  }
}
async function readStream(readable) {
  if (!readable) {
    return Buffer.from(new Uint8Array(0));
  }
  return new Promise((resolve, reject) => {
    const chunks = [];
    readable.on("error", reject);
    readable.on("data", (data) => {
      chunks.push(data);
    });
    readable.on("close", () => resolve(Buffer.concat(chunks)));
  });
}

var LogLevel;
(function(LogLevel2) {
  LogLevel2["TRACE"] = "TRACE";
  LogLevel2["DEBUG"] = "DEBUG";
  LogLevel2["INFO"] = "INFO";
  LogLevel2["WARN"] = "WARN";
  LogLevel2["ERROR"] = "ERROR";
})(LogLevel || (LogLevel = {}));
var LOG_LEVELS = Object.values(LogLevel);

var _BaseCachingSecret_value;
_BaseCachingSecret_value = /* @__PURE__ */ new WeakMap();

globalThis.getEventualHook ??= () => {
  throw new Error("Eventual Hook is not yet registered");
};
globalThis.tryGetEventualHook ??= () => {
  return void 0;
};

var ResultKind;
(function(ResultKind2) {
  ResultKind2[ResultKind2["Resolved"] = 0] = "Resolved";
  ResultKind2[ResultKind2["Failed"] = 1] = "Failed";
})(ResultKind || (ResultKind = {}));

globalThis.getEventualTaskRuntimeContext ??= () => {
  throw new Error("Eventual task context has not been registered yet.");
};

var myHandler = api.get("/", async () => {
  return new HttpResponse();
});
var myHandler2 = api.get("/", async () => {
  return new HttpResponse();
});
0 && (module.exports = {
  myHandler,
  myHandler2
});
"
`;
